### PR TITLE
feat: Lock updates on azure resources when other component is doing t…

### PR DESCRIPTION
**Pull request by nilo19**
…he same thing.

This PR utilizes a lease in each service reconciliation to prevent race conditions where cloud provider and others are updating the same azure resources.

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR utilizes a lease in each service reconciliation to prevent race conditions where cloud provider and others are updating the same azure resources. If the lease has not expired yet and is held by other component, the service reconciliation will be aborted and retried exponentially.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
feat: Lock updates on azure resources when other component is doing the same thing.

This PR utilizes a lease in each service reconciliation to prevent race conditions where cloud provider and others are updating the same azure resources.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by zarvd** [comment]
I’m curious about which component would access the same Azure resource simultaneously, and exactly which resource is it? It seems like the lock covers the entire `EnsureLoadBalancer`.

File: pkg/provider/lock/lock.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+)
+
+type AzureResourceLocker struct {
+	clientset            kubernetes.Interface
+	holder               string
+	leaseName            string
+	leaseNamespace       string
+	leaseDurationSeconds int32
```
[the remark this answers is not in the document]
Comment (nilo19) [reply pkg/provider/lock/lock.go:38]: Need an int32 for lease api.

[the remark this answers is not in the document]
Comment (zarvd) [reply pkg/provider/lock/lock.go:38]: you can still convert it to int32 for internal implementation

[the remark this answers is not in the document]
Comment (nilo19) [reply pkg/provider/lock/lock.go:38]: What would be the benefit of adding an extra conversion?

[the remark this answers is not in the document]
Comment (zarvd) [reply pkg/provider/lock/lock.go:38]: Oh, I think it’s a comment on the `NewAzureResourceLocker`. For that, I just think using a type or abstraction would be more idiomatic. Just a nitpick.

[the remark this answers is not in the document]
Comment (nilo19) [reply pkg/provider/lock/lock.go:38]: Understand, but to me personally, something "duration in seconds" indicates that it is not a duration but a number representing seconds.

File: pkg/provider/azure_lock.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+	}
+
+	// invalidate caches if the previous holder was not the cloud controller manager
+	// TODO: invalidates more caches if needed
+	if lease.Annotations != nil &&
+		!strings.EqualFold(
+			lease.Annotations[consts.AzureResourceLockPreviousHolderNameAnnotation],
+			consts.AzureResourceLockHolderNameCloudControllerManager,
+		) {
+		l.Cloud.lbCache, err = l.Cloud.newLBCache()
```
Comment (nilo19) [inline pkg/provider/azure_lock.go:171]: Invalidates cache if the previous holder is another component. We can invalidate more caches here if needed.

Comment (feiskyer) [reply pkg/provider/azure_lock.go:172]: we'd need clear caches for vmss, vmssvm and vm

Comment (nilo19) [reply pkg/provider/azure_lock.go]: Added.

**Review by feiskyer — APPROVED** [review]
/lgtm

**Comment by nilo19** [comment]
E1012 01:41:09.094776       1 controllermanager.go:93] Run: failed to configure cloud controller manager: --cloud-config cannot be empty when --enable-dynamic-reloading is not set to true

**Pull request merged**